spec/integration/slices_spec.rb

```diff
@@ -81,6 +81,188 @@ class Slice < Hanami::Slice
+        require "hanami"
+
+        module TestApp
+          class App < Hanami::App
+          end
+        end
+      RUBY
+
+      write "slices/main/config/slice.rb", <<~RUBY
+        module Main
+          class Slice < Hanami::Slice
+            config.actions.format :json
+          end
+        end
+      RUBY
+
+      require "hanami/prepare"
+
+      expect(Hanami.app.slices[:main]).to be Main::Slice
+      expect(Hanami.app.slices[:main].config.actions.format).to eq([:json])
+    end
+  end
+
+  it "Loading a slice with its own defined slice class prefers the app's defined slice class" do
+    with_tmp_directory(Dir.mktmpdir) do
+      write "config/app.rb", <<~RUBY
+        require "hanami"
+
+        module TestApp
+          class App < Hanami::App
+          end
+        end
+      RUBY
+
+      write "config/slices/main.rb", <<~RUBY
+        require "hanami"
+
+        module Main
+          class Slice < Hanami::Slice
+            config.actions.format :html
+          end
+        end
+      RUBY
+
+      write "slices/main/config/slice.rb", <<~RUBY
+        module Main
+          class Slice < Hanami::Slice
+            config.actions.format :json
+          end
+        end
+      RUBY
+
+      require "hanami/prepare"
+
+      expect(Hanami.app.slices[:main]).to be Main::Slice
+      expect(Hanami.app.slices[:main].config.actions.format).to eq([:html])
+    end
+  end
+
+  it "Loading a nested slice with its own defined slice class" do
+    with_tmp_directory(Dir.mktmpdir) do
+      write "config/app.rb", <<~RUBY
+        require "hanami"
+
+        module TestApp
+          class App < Hanami::App
+          end
+        end
+      RUBY
+
+      write "slices/main/slices/nested/config/slice.rb", <<~RUBY
+        module Main
+          module Nested
+            class Slice < Hanami::Slice
+              config.actions.format :json
+            end
+          end
+        end
+      RUBY
+
+      require "hanami/prepare"
+
+      expect(Hanami.app.slices[:main]).to be Main::Slice
+      expect(Hanami.app.slices[:main].slices[:nested].config.actions.format).to eq([:json])
+    end
+  end
+
+  it "Loading a nested slice with its own defined slice class prefers the parent slice's defined slice class" do
+    with_tmp_directory(Dir.mktmpdir) do
+      write "config/app.rb", <<~RUBY
+        require "hanami"
+
+        module TestApp
+          class App < Hanami::App
+          end
+        end
+      RUBY
+
+      write "slices/main/config/slices/nested.rb", <<~RUBY
+        require "hanami"
+
+        module Main
+          module Nested
+            class Slice < Hanami::Slice
+              config.actions.format :html
+            end
+          end
+        end
+      RUBY
+
+      write "slices/main/slices/nested/config/slice.rb", <<~RUBY
+        module Main
+          module Nested
+            class Slice < Hanami::Slice
+              config.actions.format :json
+            end
+          end
+        end
+      RUBY
```

Comment:
Likewise, let's replace this with a simple `raise` if you agree with my thoughts on the earlier example. After seeing this a second time (in this much more complex nested structure), I think this approach can really improve understanding of the test.

Reply:
(Same for the test below, but I won't bother you with another inline comment about that.)

Reply:
works great!